### PR TITLE
[KEYCLOAK-7680] Include the RH-SSO 7.3 TP CD adapters into EAP CD image

### PR DESCRIPTION
**Pull request by iankko**
https://issues.jboss.org/browse/CLOUD-2910
Signed-off-by: Jan Lieskovsky <jlieskov@redhat.com>

Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:

- [ ] Pull Request title is properly formatted: `[CLOUD-XYA] Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Attached commits represent units of work and are properly formatted
- [ ] You have read and agreed to the Developer Certificate of Origin (DCO) (see `CONTRIBUTING.md`)
- [ ] Every commit contains `Signed-off-by: Your Name <yourname@example.com>` - use `git commit -s`

### CODE REVIEW

**Comment by iankko** [comment]
@rcernich @ASzc @luck3y PTAL (once got a chance)

Thank you, Jan

**Review by luck3y — DISMISSED** [review]
Looks good to me, thanks Jan!

I'll do an equivalent change for EAP CD.

Edit never mind, I see this is going to eap-cd-dev :) I meant I'll cherry pick it into CD 14.

**Comment by iankko** [comment]
@luck3y Yes, this needs to be done for ```eap-cd``` branch too. Thanks!

File: image.yaml [hunk 1]
```diff
@@ -56,13 +56,13 @@ modules:
           - name: os-eap-probes
           - name: jboss-maven
           - name: os-eap-db-drivers
-          - name: os-eap-sso
           - name: os-eap70-sso
```
Comment (luck3y) [inline image.yaml]: BTW, @iankko is keeping os-eap70-sso here going to be a problem, or should it be moved too?

Comment (iankko) [reply image.yaml]: AFAICT it can be moved too as far as it is placed right before the call to ```os-eap-extensions``` module (so the default adapters version are installed a bit sooner than the possibly overridden ones from the extension module). Want me to move ```os-eap70-sso``` below too?

**Comment by iankko** [comment]
@maschmid @Crumby PTAL (once got a chance)

Thank you, Jan

File: image.yaml (hunk 1 above)
Comment (luck3y) [reply image.yaml]: Moving it would seem to make sense to me, yeah, since the eap70 one is installing adapters too.

Comment (iankko) [reply image.yaml]: Ok, will move it then above the extensions module (so it still works for KEYCLOAK-7680 purpose).

**Comment by iankko** [comment]
@luck3y Moved. Also added a link to relevant JIRA, so in the future we know, why the order matters :)

File: image.yaml [hunk 2]
```diff
@@ -56,13 +56,14 @@ modules:
           - name: os-eap-probes
           - name: jboss-maven
           - name: os-eap-db-drivers
-          - name: os-eap-sso
-          - name: os-eap70-sso
           - name: os-java-hawkular
           - name: os-eap70-hawkular
           - name: os-eap-hawkular
           - name: os-eap-deployment-scanner
+          - name: os-eap70-sso
```
Comment (rcernich) [inline image.yaml]: Are these changes necessary?  All they do is change the order in which the two modules are installed.

File: templates/eap-cd-sso-s2i.json [hunk 3]
```diff
@@ -489,6 +504,10 @@
                             "name": "eap-cd-openshift:12"
                         },
                         "env": [
+                            {
+                                "name": "CUSTOM_INSTALL_DIRECTORIES",
+                                "value": "rh-sso/client_adapters/*"
```
Comment (rcernich) [inline templates/eap-cd-sso-s2i.json]: I think we should leave this as `extensions/*` and the mount point for the sso adapters should be changed to `"destinationDir": "./${CONTEXT_DIR}/extensions/sso-adapters`.  The trick here is that the glob will scan all directory children of extensions, which would allow other extensions to be added without modifying `CUSTOM_INSTALL_DIRECTORIES`, e.g. adding a db driver image mounted at `./${CONTEXT_DIR}/extensions/some-db`.  We could even consider specifying `extensions/*` as the default setting, which would allow users to just add image sources without having to do other modifications (e.g. adding a db driver after using the _basic_ templates).

**Review by rcernich — CHANGES_REQUESTED** [review]

**Comment by luck3y** [comment]
The issue we have with this PR is that CD 12 is finished, to be replaced with CD 13, which is past feature changes, and being tested by QE. So I don't think we can merge this into CD 13.

File: image.yaml (hunk 2 above)
Comment (iankko) [reply image.yaml]: This specific one isn't (performed it just upon Ken's request). But the ```os-eap-sso``` one (to be called after ```os-eap-extensions```) is (so I wouldn't need to call the RH-SSO / Keycloak adapters configuration two times.

File: templates/eap-cd-sso-s2i.json (hunk 3 above)
Comment (iankko) [reply templates/eap-cd-sso-s2i.json]: Ok. Thank you for the suggestions, fair enough. Will perform those changes.

**Comment by iankko** [comment]
@rcernich PR updated with your suggestions! Thank you for those.

**Comment by iankko** [comment]
@rcernich Actually, after testing the changed ```destinationDir``` and ```CUSTOM_INSTALL_DIRECTORIES``` values (the suggestions), there's something wonky with them. The adapters aren't expanded as part of the EAP S2I image build process (there's no ```Inflating ....```) message in the pod log. And the EAP pod when running contains only default RH-SSO adapters versions.

On the other hand, the ```redhat-sso-7-image``` change (checked also the updated one from Alex) works properly with the old form of both the ```destinationDir``` and the ```CUSTOM_INSTALL_DIRECTORIES``` values. In that case the custom adapters are properly expanded (within EAP S2I image build process), and the EAP pod contains just the customized versions.

From a brief look IDK what's the difference, causing this behaviour. But the reason doesn't seem to be the ```redhat-sso-7``` image change (since it works with old values), but the template settings.

PTAL

File: templates/eap-cd-sso-s2i.json [hunk 4]
```diff
@@ -477,7 +477,22 @@
...
+                        {
+                            "from": {
+                                "kind": "ImageStreamTag",
+                                "namespace": "${IMAGE_STREAM_NAMESPACE}",
+                                "name": "redhat-sso-cd-openshift:1.0"
+                            },
+                            "paths": [
+                                {
+                                    "destinationDir": "./${CONTEXT_DIR}/extensions/sso-adapters",
+                                    "sourcePath": "/opt/rh/rh-sso/client/eap7"
```
Comment (rcernich) [inline templates/eap-cd-sso-s2i.json]: I think the issue is that this needs to end with `/.` (slash dot, i.e. `/opt/rh/rh-sso/client/eap7/.`), which copies the contents of the directory into the target location.  See here: https://docs.openshift.com/container-platform/3.9/dev_guide/builds/build_inputs.html#image-source

**Comment by iankko** [comment]
@rcernich Yes, that was it! Now it works:

inflating

And the corresponding pod contains expected adapter version(s):

updated_adapters

Thanks!

This one is ready too.

**Comment by iankko** [comment]
@luck3y @rcernich Rebased the PR against current `eap-cd-dev` content, and verified the PR still works properly. Should be ready to go. PTAL

Thanks, Jan

**Review by luck3y — APPROVED** [review]

**Comment by luck3y** [comment]
LGTM, thanks @iankko!

**Review by rcernich — APPROVED** [review]
